cynghanedd.py

```diff
@@ -205,14 +216,57 @@ def choose_word(solutions, word):
 
     words = []
 
-    while solution and solution[0]:
+    words, solution = complete_cyng(words, skeleton, solution)
+
+    return words, skeleton, solution
+
+
+def back_cynghanedd(old_words, skeleton):
+    solution = search(skeleton)
+
+    print(old_words[:-1])
+    words = []
+    i = 0
+    current_iteration = len(old_words)
+    print("Current iteration is ", end="")
+    print(current_iteration)
+
+    while i < (current_iteration - 1):
         next_words = ' '.join(sorted(first_words(solution)))
-        print(next_words)
-        print("Pick a word: ", end="", flush=True)
-        word = sys.stdin.readline().strip()
+        word = old_words[i]
+        i += 1
         words.append(word)
-
         solution = choose_word(solution, word)
+    next_words = ' '.join(sorted(first_words(solution)))
+    print(next_words)
+    print("Pick a word: ", end="", flush=True)
+    word = sys.stdin.readline().strip()
+
+    return word, words, solution
 
-    print(' '.join(words))
 
+if __name__ == '__main__':
+    import sys
+
+    words, skeleton, solution = run_cynghanedd()
+    
+    answer = ""
+    while not (answer == "n"):
```

Comment:
I think I'd probably put all of the logic related to reading a line from the user & figuring out what to do with it in this block, so that the other functions don't have to ever think about reading stuff from the user, or printing things, they can just get on with the computations.

If we structure things that way, we can have a loop hear that on each iteration prompts the user for some input & deals with just that line of input, then the next iteration of the loop handles the next line of input etc. That might end up looking something like this:

```
skeleton = None
solution = None
words = []

while True:
  if not skeleton:
    # We're right at the start of the process: either the user hasn't given us a skeleton at all, or we've just restarted
    print("Give me a skeleton")
    skeleton = sys.stdin.readline().strip().split()
    solution = search(skeleton)
  else if solution and solution[0]:
    # We're in the middle of filling in a solution.
    # ... print all the words they can choose
    word = sys.stdin.readline().strip()

    if word == ',':
      # the user typed ",", which means they want to go back a word
      words, solution = back_cynghanedd(skeleton, words)
    else:
      words, solution = choose_word(solution, word)
  else:
    # We've completed a line, ask the user if they want to start again...
    print("restart?")
    if sys.stdin.readline().strip() == 'y':
      # set all these variables back to what their values were at the beginning of the program, so that on our next iteration
      # of the loop we'll start the whole process from scratch.
      skeleton = None
      solution = None
      words = []
    else:
      # the user doesn't want to keep writing, leave the loop
      break
```

This pattern is using an idea called "purity": basically code that interacts with the outside world (e.g. the user) is called "impure", while code that just does computation is called "pure". In general, things are much easier to deal with when you keep the pure and impure parts separate: in the example I just gave, all the impure stuff (the `print` and `sys.stdin.readline()` lines) is in the body of the big loop; all the other functions are pure.